lib/Value/Matrix.pm

```diff
@@ -493,6 +493,110 @@ sub I {
 	return $self->make($context, @M);
 }
 
+#
+#  Get an elementary matrix of the requested size and type
+#  E(n,[i,j])   nxn, swap rows i and j
+#  E(n,[i,j],k) nxn, replace row i with row i added to k times row j
+#  E(n,[i],k)   nxn, scale row i by k
+#
+sub E {
+	my ($self, $d, $rows, $k, $context) = @_;
+	$context = $self->context unless $context;
+	Value::Error("You must provide a dimension for an Elementary matrix") unless defined $d;
+	Value::Error("Dimension must be a positive integer")                  unless $d =~ m/^[1-9]\d*$/;
+	my @ij = @{$rows};
+	Value::Error("Either one or two rows must be specified for an Elementary matrix") unless (@ij == 1 || @ij == 2);
+	Value::Error(
+		"If only one row is specified for an Elementary matrix, then a number to scale by must also be specified")
+		if (@ij == 1 && !defined $k);
+	for (@ij) {
+		Value::Error("Row numbers must be integers between 1 and $d")
+			unless ($_ =~ m/^[1-9]\d*$/ && $_ >= 1 && $_ <= $d);
+	}
+	@ij = map { $_ - 1 } (@ij);
+
+	my @M    = ();
+	my @Z    = split('', 0 x $d);
+	my $REAL = $context->Package('Real');
+
+	foreach my $i (0 .. $d - 1) {
```

Comment:
I think for consistency, we're trying to move foreach over to for. There are a few places here to do that.